component/remote/vault/refresher.go

```diff
@@ -0,0 +1,278 @@
+	token := tm.token
+	tm.mut.RUnlock()
+
+	tm.debugMut.Lock()
+	defer tm.debugMut.Unlock()
+
+	tm.debugInfo = getSecretInfo(token, updateTime)
+}
+
+func (tm *tokenManager) updateLifecycleWatcher(ctx context.Context) {
+	tm.mut.RLock()
+	defer tm.mut.RUnlock()
+
+	if !needsLifecycleWatcher(tm.token) {
+		return
+	}
+
+	lw, err := tm.cli.NewLifetimeWatcher(&vault.LifetimeWatcherInput{
+		Secret:        tm.token,
+		RenewBehavior: vault.RenewBehaviorIgnoreErrors,
+	})
+	if err != nil {
+		level.Error(tm.log).Log("msg", "failed to create lifetime watcher, lease will not renew automatically", "err", err)
+		return
+	}
+
+	go lw.Start()
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				lw.Stop()
+				return
+
+			case <-lw.DoneCh():
+				if ctx.Err() != nil {
+					return
+				}
+				tm.updateToken(ctx)
+
+			case output := <-lw.RenewCh():
+				tm.refreshCounter.Inc()
+				level.Debug(tm.log).Log("msg", "token has renewed")
+				tm.updateDebugInfo(output.RenewedAt)
+			}
+		}
+	}()
+}
+
+// needsLifecycleWatcher determines if a secret needs a lifecycle watcher.
+// Secrets only need a lifecycle watcher if they are renewable or have a lease
+// duration.
+func needsLifecycleWatcher(secret *vault.Secret) bool {
+	if secret.Auth != nil {
+		return secret.Auth.Renewable || secret.Auth.LeaseDuration > 0
+	}
+	return secret.Renewable || secret.LeaseDuration > 0
+}
+
+// SetClient updates the client associated with the tokenManager. This will
+// force a new retrieval of the token.
+func (tm *tokenManager) SetClient(cli *vault.Client) {
+	tm.mut.Lock()
+	tm.cli = cli
+	tm.mut.Unlock()
+
+	ctx, cancel := context.WithTimeout(context.Background(), tokenManagerInitializeTimeout)
+	defer cancel()
+
+	tm.updateToken(ctx)
+}
+
+// SetRefrehInterval sets a forced refresh interval, separate from automatic
+// renewal based on the token lease.
+func (tm *tokenManager) SetRefreshInterval(interval time.Duration) {
+	tm.refreshTicker.Reset(interval)
+}
+
+// CurrentHealth returns the health of the tokenManager.
+func (tm *tokenManager) CurrentHealth() component.Health {
+	tm.healthMut.RLock()
+	defer tm.healthMut.RUnlock()
+
+	return tm.health
+}
+
+// DebugInfo returns the current DebugInfo for the tokenManager.
+func (tm *tokenManager) DebugInfo() secretInfo {
+	tm.debugMut.RLock()
+	defer tm.debugMut.RUnlock()
+
+	return tm.debugInfo
+}
+
+type secretInfo struct {
+	LatestRequestID  string    `river:"latest_request_id,attr"`
+	LastUpdateTime   time.Time `river:"last_update_time,attr"`
+	SecretExpireTime time.Time `river:"secret_expire_time,attr"`
+	Renewable        bool      `river:"renewable,attr"`
+	Warnings         []string  `river:"warnings,attr"`
+}
+
+func getSecretInfo(secret *vault.Secret, updateTime time.Time) secretInfo {
+	return secretInfo{
+		LatestRequestID:  secret.RequestID,
+		LastUpdateTime:   updateTime,
+		SecretExpireTime: secretExpireTime(secret),
+		Renewable:        secret.Renewable,
+		Warnings:         secret.Warnings,
+	}
+}
```

Comment:
This can panic in cases where a nil secret is returned (eg. from an authManager using the AuthToken method to authenticate).